metashare/api/jobs.py

```diff
@@ -0,0 +1,151 @@
+import contextlib
+import logging
+import traceback
+
+from asgiref.sync import async_to_sync
+from channels.layers import get_channel_layer
+from cumulusci.core.config import BaseProjectConfig, ScratchOrgConfig
+from cumulusci.core.runtime import BaseCumulusCI
+from django.utils.text import slugify
+from django.utils.timezone import now
+from django_rq import job
+from github3 import login
+from github3.exceptions import UnprocessableEntity
+
+from .github_context import (
+    extract_owner_and_repo,
+    get_cumulus_prefix,
+    local_github_checkout,
+)
+
+logger = logging.getLogger(__name__)
+
+
+def report_scratch_org_error(instance, message):
+    from .serializers import ScratchOrgSerializer
+
+    message = {
+        "type": "SCRATCH_ORG_PROVISION_FAILED",
+        "payload": {"message": message, "model": ScratchOrgSerializer(instance).data},
+    }
+    model_name = instance._meta.model_name
+    id = str(instance.id)
+    group_name = "{model}.{id}".format(model=model_name, id=id)
+    channel_layer = get_channel_layer()
+    sent_message = {"type": "notify", "group": group_name, "content": message}
+    async_to_sync(channel_layer.group_send(group_name, sent_message))
+
+
+@contextlib.contextmanager
+def report_errors_on(scratch_org):
+    try:
+        yield
+    except Exception as e:
+        report_scratch_org_error(scratch_org, str(e))
+        tb = traceback.format_exc()
+        logger.error(tb)
+        raise
+
+
+class MetadeployProjectConfig(BaseProjectConfig):
+    def __init__(
+        self, *args, repo_root=None, repo_url=None, commit_ish=None, **kwargs
+    ):  # pragma: nocover
+
+        owner, repo_name = extract_owner_and_repo(repo_url)
+
+        repo_info = {
+            "root": repo_root,
+            "url": repo_url,
+            "name": repo_name,
+            "owner": owner,
+            "commit": commit_ish,
+        }
+
+        super().__init__(*args, repo_info=repo_info, **kwargs)
+
+
+class MetaDeployCCI(BaseCumulusCI):
+    project_config_class = MetadeployProjectConfig
+
+
+def create_scratch_org(*, scratch_org, user, repo_url, commit_ish):
+    """
+    Expects to be called in the context of a local github checkout.
+    """
+    gh = login(token=user.gh_token)
+    owner, repo = extract_owner_and_repo(repo_url)
+    repository = gh.repository(owner, repo)
+    branch = repository.branch(commit_ish)
+    latest_commit = branch.latest_sha()
+    # We can't use the urlt objects in repository because they build API urls:
+    latest_commit_url = f"{repository.html_url}/commit/{latest_commit}"
+
+    with report_errors_on(scratch_org):
+        org = ScratchOrgConfig({"config_file": "orgs/dev.json", "scratch": True}, "dev")
```

Comment:
@wlonk Should this depend upon the POSTed `org_type` instead of hard-coding `dev`?

Reply:
The `dev` org config should generally be fine for both dev and QA use cases.